src/Garn/CodeGen.hs

```diff
@@ -19,69 +24,111 @@ import Data.String.Interpolate (i)
 import Data.String.Interpolate.Util (unindent)
 import GHC.Generics (Generic)
 import Garn.Common (currentSystem, nixpkgsInput)
+import System.Directory (createDirectoryIfMissing, removeDirectoryRecursive)
 import WithCli (withCli)
 
+-- A specification for what packages to include in the final collection
+-- collectionSpec represents a colleciton with all derivations included by
+-- default, but everything else excluded. This can be overridden with the
+-- attribute set.
+pkgSpec :: String
+pkgSpec =
+  [i|
+    collectionSpec {
+      haskellPackages = collectionSpec {};
+      nimPackages = collectionSpec {};
+      nodePackages = collectionSpec { "@antora/cli" = false; };
+      phpPackages = collectionSpec {};
+      python2Packages = collectionSpec {};
+      python3Packages = collectionSpec {};
+      rubyPackages = collectionSpec {};
+      rustPackages = collectionSpec {};
+    }
+  |]
+
+pkgs :: String -> String
+pkgs system =
+  [i|
+    import (builtins.getFlake "#{nixpkgsInput}") {
+      system = "#{system}";
+      config.allowAliases = false;
+    }
+  |]
+
 run :: IO ()
 run = withCli $ do
   system <- currentSystem
-  let varName = "pkgs"
-      nixpkgsExpression =
-        [i|
-          import (builtins.getFlake "#{nixpkgsInput}") {
-            system = "#{system}";
-            config.allowAliases = false;
-          }
-        |]
-  code <- fromToplevelDerivation "." varName nixpkgsExpression
-  writeFile "ts/nixpkgs.ts" code
-
-fromToplevelDerivation :: String -> String -> String -> IO String
-fromToplevelDerivation garnLibRoot varName rootExpr = do
-  system :: String <- do
-    StdoutUntrimmed json <- Cradle.run "nix" nixArgs (words "eval --impure --json --expr builtins.currentSystem")
-    pure $ either error id $ eitherDecode (cs json)
+  let outDir = "ts/internal/nixpkgs"
+  removeDirectoryRecursive outDir `catch` \(_ :: IOException) -> pure ()
+  pkgs <- scanPackages system (pkgs system) pkgSpec
+  writePkgFiles outDir "../.." pkgs
+
+writePkgFiles :: String -> String -> Map String PkgInfo -> IO ()
+writePkgFiles modulePath garnLibRoot pkgs = do
+  createDirectoryIfMissing True modulePath
+  let sanitizedPkgs = Map.mapKeys sanitize pkgs
```

Comment:
```suggestion
writePkgFiles modulePath garnLibRoot (Map.mapKeys sanitize -> pkgs) = do
  createDirectoryIfMissing True modulePath
```

With https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/view_patterns.html. That way we don't have the unsanitized `pkgs` in scope.